src/Workspaces/Core/Portable/TemporaryStorage/TemporaryStorageServiceFactory.cs

```diff
@@ -101,14 +101,22 @@ public ITemporaryTextStorageInternal CreateTemporaryTextStorage()
         string storageName, long offset, long size, SourceHashAlgorithm checksumAlgorithm, Encoding? encoding, ImmutableArray<byte> contentHash)
         => new(this, storageName, offset, size, checksumAlgorithm, encoding, contentHash);
 
-    ITemporaryStreamStorageInternal ITemporaryStorageServiceInternal.CreateTemporaryStreamStorage()
-        => CreateTemporaryStreamStorage();
+    public TemporaryStorageHandle WriteToTemporaryStorage(Stream stream, CancellationToken cancellationToken)
+    {
+        var storage = new TemporaryStreamStorage(this);
+        storage.WriteStream(stream, cancellationToken);
+        var identifier = new TemporaryStorageIdentifier(storage.Name, storage.Offset, storage.Size);
+        return new(storage, identifier);
+    }
 
-    internal TemporaryStreamStorage CreateTemporaryStreamStorage()
-        => new(this);
+    Stream ITemporaryStorageServiceInternal.ReadFromTemporaryStorageService(TemporaryStorageIdentifier storageIdentifier, CancellationToken cancellationToken)
+        => ReadFromTemporaryStorageService(storageIdentifier, cancellationToken);
```

Comment:
the interface exposes a weakly types "stream-returning" api.  The concrete type returns an UnmanagedMemoryStream , which makes a lot of consumption side code much simpler, knowing when the data is backed by pinned, unmanaged, memory.